parser/src/lexer.rs

```diff
@@ -214,299 +131,259 @@ pub fn lex_starts_at(
-                self.next_char();
-                self.next_char();
-                self.lex_number_radix(start_pos, 16)
-            }
-            [Some('0'), Some('o' | 'O')] => {
-                // Octal style! (0o377)
-                self.next_char();
-                self.next_char();
-                self.lex_number_radix(start_pos, 8)
-            }
-            [Some('0'), Some('b' | 'B')] => {
-                // Binary! (0b_1110_0101)
-                self.next_char();
-                self.next_char();
-                self.lex_number_radix(start_pos, 2)
-            }
-            _ => self.lex_normal_number(),
+    fn lex_number(&mut self, first: char) -> Result<Tok, LexicalError> {
+        if first == '0' {
+            if self.cursor.eat_if(|c| matches!(c, 'x' | 'X')).is_some() {
+                self.lex_number_radix(first, Radix::Hex)
+            } else if self.cursor.eat_if(|c| matches!(c, 'o' | 'O')).is_some() {
+                self.lex_number_radix(first, Radix::Octal)
+            } else if self.cursor.eat_if(|c| matches!(c, 'b' | 'B')).is_some() {
+                self.lex_number_radix(first, Radix::Binary)
+            } else {
+                self.lex_decimal_number(first)
+            }
+        } else {
+            self.lex_decimal_number(first)
         }
     }
 
     /// Lex a hex/octal/decimal/binary number without a decimal point.
-    fn lex_number_radix(&mut self, start_pos: TextSize, radix: u32) -> LexResult {
-        let value_text = self.radix_run(radix);
-        let end_pos = self.get_pos();
-        let value = BigInt::from_str_radix(&value_text, radix).map_err(|e| LexicalError {
-            error: LexicalErrorType::OtherError(format!("{e:?}")),
-            location: start_pos,
-        })?;
-        Ok((Tok::Int { value }, TextRange::new(start_pos, end_pos)))
+    fn lex_number_radix(&mut self, first: char, radix: Radix) -> Result<Tok, LexicalError> {
+        #[cfg(debug_assertions)]
+        debug_assert!(matches!(
+            self.cursor.previous().to_ascii_lowercase(),
+            'x' | 'o' | 'b'
+        ));
+
+        let value_text = self.radix_run(Some(first), radix);
+        let value =
+            BigInt::from_str_radix(&value_text, radix.as_u32()).map_err(|e| LexicalError {
+                error: LexicalErrorType::OtherError(format!("{e:?}")),
+                location: self.token_range().start(),
+            })?;
+        Ok(Tok::Int { value })
     }
 
     /// Lex a normal number, that is, no octal, hex or binary number.
-    fn lex_normal_number(&mut self) -> LexResult {
-        let start_pos = self.get_pos();
-        let start_is_zero = self.window[0] == Some('0');
-        // Normal number:
-        let mut value_text = self.radix_run(10);
+    fn lex_decimal_number(&mut self, first_digit_or_dot: char) -> Result<Tok, LexicalError> {
+        #[cfg(debug_assertions)]
+        debug_assert!(self.cursor.previous().is_ascii_digit() || self.cursor.previous() == '.');
```

Comment:
Yeah. I wanted to always consume the first character in next_token because this guarantees us that the lexer is always progressing (no infinite loops) 

But the fact that we don't lex whitespace makes this a bit awkward because there are now cases where we consume a character without returning a token